builder/dockerfile/internals.go

```diff
@@ -302,12 +301,13 @@ func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression
 
 	// Work in daemon-specific OS filepath semantics
 	origPath = filepath.FromSlash(origPath)
-
 	// validate windows paths from other images
 	if contextID != nil && runtime.GOOS == "windows" {
-		forbid := regexp.MustCompile("(?i)^" + string(os.PathSeparator) + "?(windows(" + string(os.PathSeparator) + ".+)?)?$")
-		if p := filepath.Clean(origPath); p == "." || forbid.MatchString(p) {
-			return nil, errors.Errorf("copy from %s is not allowed on windows", origPath)
+		p := strings.ToLower(filepath.Clean(origPath))
+
+		if p == "." || p == ".." || p == "\\" || p == "\\windows" || p == "windows" ||
+			p == "c:" || p == "c:." || p == "c:\\" || p == "c:\\windows" || p == "c:\\windows\\" {
```

Comment:
Do we want to blacklist _exactly_ `c:\windows\` or also, e.g. `c:\windows\system32`? If the latter, we should match on _prefix_ (or another way to check if `p` is _inside_ "forbiddenpath".

What about symlinks?